### PR TITLE
etcd-tester: add retry logic on retriving lease info

### LINKED ISSUE #6754: functional-tester: retrieving lease error
The 3 nodes functional-tester experienced a failure when retrieving lease info failed at lease checking phase.

```
#tester.log
135635 2016-10-28 18:12:30.444107 I | etcd-tester: [round#282 case#4] lease stresser encounters error: (rpc error: code = 2 desc = Post http://10.240.0.3:2380/leases/internal: EOF)
135636 2016-10-28 18:12:30.444127 W | etcd-tester: %!s(func() string=0x4189d0) functional-tester returning with error (tt.checkConsistency error (lease stresser encounters error: (rpc error: code = 2 de       sc = Post http://10.240.0.3:2380/leases/internal: EOF)))
135637 2016-10-28 18:12:30.444142 I | etcd-tester: [round#282 case#4] cleaning up...
```

logs can be found at https://console.cloud.google.com/storage/browser/etcd-functional-tester/lease-retriving-error-20161028/

### PR DESCRIPTION
**Pull request by fanminshi**
getting lease and keys info through raw rpcs rarely experience error such as EOF. This is considered as a failure and causes tester to clean up.
however, they are just transient problem with temporary connection issue which should not br considered as a testing failure. so we add retry logic in case of transient failure.

FIX #6754

### CODE REVIEW

**Comment by fanminshi** [comment]
I suspect the EOF issue has nothing to do with our internal server code but it is a issue with the http library in which we are using. I think the simplest way to get around it is through retrying on client side. since connection errors are expected and client should take care of those.

cc @xiang90 @gyuho @heyitsanthony

**Comment by heyitsanthony** [comment]
If it's errors from a disrupted connection, why not issue the RPC calls with FailFast=false?

**Comment by fanminshi** [comment]
@heyitsanthony 
getting lease info is using failfast=false. we are still experiencing EOF. 

```
func (ls *leaseStresser) getLeaseByID(ctx context.Context, leaseID int64) (*pb.LeaseTimeToLiveResponse, error) {
	ltl := &pb.LeaseTimeToLiveRequest{ID: leaseID, Keys: true}
	return ls.lc.LeaseTimeToLive(ctx, ltl, grpc.FailFast(false))
}
```

**Comment by heyitsanthony** [comment]
@fanminshi ok, that's a server-side issue.  There needs to be an actual error code returned by the server that can be detected here instead of dropping the error on the floor.

**Comment by fanminshi** [comment]
@heyitsanthony I dig into the lease code a bit. I think error is thrown from this line.
https://github.com/coreos/etcd/blob/master/lease/leasehttp/http.go#L187
I suspect that the http req failed due to EOF, and then that error is escalated to the client side. 
I wonder how would the server side handles this kind of connection related error?

**Comment by fanminshi** [comment]
The syndrome is similar to that of http://stackoverflow.com/questions/17714494/golang-http-request-results-in-eof-errors-when-making-multiple-requests-successi.

However, according to the post, that issue is patched in the go version 1.6. So we might experiencing something a bit different.

File: tools/functional-tester/etcd-tester/lease_stresser.go [hunk 1]
```diff
@@ -258,28 +258,35 @@ func (ls *leaseStresser) getLeaseByID(ctx context.Context, leaseID int64) (*pb.L
 }
 
 func (ls *leaseStresser) hasLeaseExpired(ctx context.Context, leaseID int64) (bool, error) {
-	resp, err := ls.getLeaseByID(ctx, leaseID)
-	plog.Debugf("hasLeaseExpired %v resp %v error (%v)", leaseID, resp, err)
-	if rpctypes.Error(err) == rpctypes.ErrLeaseNotFound {
-		return true, nil
+	for ctx.Err() == nil {
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/lease_stresser.go]: this for condition seems strange. how about checking the ctx.Err at the end of the loop instead?

Comment (xiang90) [inline tools/functional-tester/etcd-tester/lease_stresser.go]: also i think we only need to retry on the keepalive path, not here?

Comment (fanminshi) [inline tools/functional-tester/etcd-tester/lease_stresser.go]: @xiang90 I wonder which keepalive path are you talking about?

Comment (xiang90) [inline tools/functional-tester/etcd-tester/lease_stresser.go]: sorry. my mistake. i was confused.

File: tools/functional-tester/etcd-tester/checks.go [hunk 2]
```diff
@@ -54,14 +54,14 @@ func (hc *hashChecker) checkRevAndHashes() (err error) {
 	for i := 0; i < retries; i++ {
 		revs, hashes, err = hc.hrg.getRevisionHash()
 		if err != nil {
-			plog.Warningf("retry %i. failed to retrieve revison and hash (%v)", i, err)
+			plog.Warningf("retry %d. failed to retrieve revison and hash (%v)", i, err)
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/checks.go]: can we fix this in another pr? thanks!

File: tools/functional-tester/etcd-tester/lease_stresser.go [hunk 3]
```diff
@@ -258,12 +258,18 @@ func (ls *leaseStresser) getLeaseByID(ctx context.Context, leaseID int64) (*pb.L
 }
 
 func (ls *leaseStresser) hasLeaseExpired(ctx context.Context, leaseID int64) (bool, error) {
-	resp, err := ls.getLeaseByID(ctx, leaseID)
-	plog.Debugf("hasLeaseExpired %v resp %v error (%v)", leaseID, resp, err)
-	if rpctypes.Error(err) == rpctypes.ErrLeaseNotFound {
-		return true, nil
+	// keep retrying until lease's state is known or ctx is being canceled
+	for ctx.Err() == nil {
+		resp, err := ls.getLeaseByID(ctx, leaseID)
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/lease_stresser.go]: the error is not expected, right? probably we should print out warning here if there is an error other than lease not found.

**Comment by xiang90** [comment]
Let's clean this up. Then lgtm

File: tools/functional-tester/etcd-tester/checks.go (hunk 2 above)
Comment (fanminshi) [inline tools/functional-tester/etcd-tester/checks.go]: yep. will do

File: tools/functional-tester/etcd-tester/lease_stresser.go (hunk 3 above)
Comment (fanminshi) [inline tools/functional-tester/etcd-tester/lease_stresser.go]: yeah, there shouldn't be error the first place. a warning makes sense.

File: tools/functional-tester/etcd-tester/lease_stresser.go [hunk 4]
```diff
@@ -258,12 +258,18 @@ func (ls *leaseStresser) getLeaseByID(ctx context.Context, leaseID int64) (*pb.L
...
+		plog.Warningf("hasLeaseExpired %v resp %v error (%v)", leaseID, resp, err)
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/lease_stresser.go]: move this line after 270. we do not want to print it out if:

1. there is no error

2. the error is not found

**Comment by fanminshi** [comment]
All fixed. PTAL @xiang90

**Comment by xiang90** [comment]
@fanminshi LGTM. Thanks a lot!

**Pull request merged**